rfcs/0028-command-line-registry-composition.md

```diff
@@ -0,0 +1,153 @@
+                        target llvm mcpu string (default: None)
+  --target-llvm-mattr TARGET_LLVM_MATTR
+                        target lvm mattr string (default: None)
+```
+
+These are arranged in per-`Target` groups to allow the user to easily follow which are applicable and which aren't. The use can now replace their existing `Target` string, by using the documented `Target` options, with:
+```
+tvmc --target=woofles,c \
+    --target-woofles-mcpu=woof \
+    --target-c-mattr=+mwoof
+```
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Alongside support for a target string, `tvmc` would populate a series of other arguments specific to targets which match the attributes found in the TargetKindRegistry in `target_kind.cc`. For example, consider a subset of the `c` `Target`:
+
+```
+TVM_REGISTER_TARGET_KIND("c", kDLCPU)
+    .add_attr_option<String>("mcpu")
+```
+
+This would be translated at the `tvmc` level to:
+```bash
+tvmc --target=c \
+    --target-c-mcpu=cortex-m3
+```
+
+Which would then allow the user to compose together these options, such as:
+
+```bash
+tvmc --target=cmsisnn,c \ # Specifying multiple targets to enable in priority order
+    --target-cmsisnn-mattr=+dsp \
+    --target-c-mcpu=cortex-m3
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+There already exists a mechanism which provides Python with `PassConfig` information, via:
+
+```c++
+// Function Registry
+TVM_REGISTER_GLOBAL("transform.ListConfigs").set_body_typed(PassContext::ListConfigs);
+// Actual call
+Map<String, Map<String, String>> PassContext::ListConfigs() {
+  return PassConfigManager::Global()->ListConfigs();
+}
+// Implementation
+Map<String, Map<String, String>> ListConfigs() {
+    Map<String, Map<String, String>> configs;
+    for (const auto& kv : key2vtype_) {
+        Map<String, String> metadata;
+        metadata.Set("type", kv.second.type_key);
+        configs.Set(kv.first, metadata);
+    }
+    return configs;
+}
+```
+
+This can be replicated to provide the same information for `TargetKind` or any other registry and provide this information to generate arguments in `tvmc` using `argparse`:
+```
+parser.add_argument(f"--target-{kind}-{attr}", ...)
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+There's a number of alternative methods for injecting this information which have been used in other CLIs but obfuscate some of the information from the CLI arguments themselves:
+- Continue using target strings
+- Specify JSON
+
+# Prior art
+[prior-art]: #prior-art
+
+This is not dissimilar to how `gcc` or `clang` provide options, [for example in gcc](https://gcc.gnu.org/onlinedocs/gcc/Instrumentation-Options.html):
+```
+gcc -fsanitize=address -fsanitize-address-use-after-scope
+```
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
```

Comment:
i think there are some unresolved questions in my mind around handling of heterogeneous targets (e.g. with >2 target), and around some efforts i think @junrushao1994 is familiar with around representing and parsing targets from JSON.

Reply:
Is there an action here @areusch ?

Reply:
sorry--was ccing @junrushao1994 as i couldn't remember where this was. there is currently `target._load_config_dict` as a way to parse Target from a JSON object. could we specify a way to work with this (or promote it to be the value of the `"target"` key here)?

Reply:
the unresolved questions around heterogeneous targets essentially are: i suppose we will modify the `load_cnfig_dict` to work with [`TargetDevice`](https://github.com/apache/tvm/pull/8892) (e.g. we will assign a name to targets and likely the way that would make sense in `load_config_dict` is to make the target dict a sort of map from `targetdevice_name` to roughly the current dict expected by `load_config_dict`. i'm raising this here only because this is user-facing and i'd like to ensure we are surfacing any internal representations as far up as possible so it's clear what may need to change as we make these structural compiler changes.

Reply:
Apologies, I'm not entirely sure what you're suggesting here, do you suggest `--target` takes JSON? Or some internal mechanism converts the arguments shown here into an intermediate representation for `target._load_config_dict` ?

Reply:
Hi, just want to mention that `load_config_dict` is a function we use to parse target json strings in both python & c++ side using python json library. It is currently designed for single target though. I think it would take some effort, while feasible, to make modifications to work with heterogenes targets if we want to make the change.